### PR TITLE
Issue #49: Preview own unpublished permission

### PR DESCRIPTION
**Pull request by donquixote**
Replaces:
- #7

### CODE REVIEW

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+                    "Kelly's published document" => ['edit in collabora'],
+                    "Kelly's unpublished document" => ['edit in collabora'],
+                ],
+                'Media admin' => [
+                    "Sean's published document" => ['preview in collabora', 'edit in collabora'],
+                    "Sean's unpublished document" => ['preview in collabora', 'edit in collabora'],
+                    "Kelly's published document" => ['preview in collabora', 'edit in collabora'],
+                    "Kelly's unpublished document" => ['preview in collabora', 'edit in collabora'],
+                ],
+            ],
```
Comment (donquixote) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:106]: I think this new order of array keys is more intuitive.

The cross product strategy causes some kind of redundancy, but that's ok because it runs fast and is not overly verbose.

File: src/CollaboraMediaPermissions.php [hunk 2]
```diff
@@ -77,7 +77,10 @@ protected function buildPermissions(MediaTypeInterface $type) {
 
         return [
             "preview $type_id in collabora" => [
-                'title' => $this->t('%type_name: Preview media file in Collabora', $type_params),
+                'title' => $this->t('%type_name: Preview published media file in Collabora', $type_params),
```
Comment (AaronGilMartinez) [inline src/CollaboraMediaPermissions.php:80]: Should we do this distinction (published)?

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+                : 'Users with access to given paths',
+        );
+    }
+
+    /**
+     * Verifies basic assumptions for this test.
+     *
+     * It is enough to call this from one test method.
+     */
+    protected function assertBasicAssumptions(): void {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: I don't think is needed to test that media doesn't add permissions to anonymous and authenticated if you don't call media_install().

Same for the id of the second user. Comments are enough.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+     * @param array<string, \Drupal\Core\Session\AccountInterface> $accounts
+     *   Accounts to test access with, keyed by a distinguishable name.
+     * @param array<string, string>|null $paths
+     *   An array of paths, or NULL to just use the array keys from $expected.
+     *   This parameter is useful if the paths all look very similar.
+     * @param bool $negate
+     *   FALSE, if $expected describes access being granted.
+     *   TRUE, if $expecte describes access being denied.
+     */
+    protected function assertPathsAccessByUsers(array $expected, array $accounts, ?array $paths = NULL, bool $negate = FALSE): void {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: I don't see this negate parameter used anywhere, I'd say if we don't need it we drop it. Same for other functions.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+                '/cool/view/%s',
+                '/cool/edit/%s',
+            ],
+        );
+    }
+
+    /**
+     * Tests a scenario where the anonymous user has more permissions.
+     */
+    public function testAnonymousOwnAccess(): void {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:157]: Users will have access based on the permissions, status and ownership, if anonymous ownership is treated different or not shouldn't be part of this test. Tests should cover what the module is implementing.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        user_role_grant_permissions(RoleInterface::ANONYMOUS_ID, [
+            'preview own unpublished document in collabora',
+            'edit own document in collabora',
+        ]);
+
+        /** @var \Drupal\Core\Session\AccountInterface[] $accounts */
+        $accounts = [
+            'anonymous' => new AnonymousUserSession(),
+            'Emilia' => $this->createUser(),
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: I think Emilia is not needed here, you're already checking that a user without permsisions doens't have access to anything in testCollaboraMediaAccess().

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+     *   Entities to check.
+     * @param array<string, \Drupal\Core\Entity\EntityInterface> $entities
+     *   Operations to check.
+     * @param list<string> $operations
+     *   User accounts to check.
+     * @param bool $negate
+     *   FALSE, if $expected describes access being granted.
+     *   TRUE, if $expecte describes access being denied.
+     */
+    protected function assertEntityAccess(array $expected, array $accounts, array $entities, array $operations, bool $negate = FALSE): void {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: Operations are always the same, no need to create a parameter for them.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+     *   Accounts to test access with, keyed by a distinguishable name.
+     * @param array<string, \Drupal\Core\Entity\EntityInterface> $entities
+     *   Entities for which to build paths.
+     * @param array<string, string> $sprintf_path_patterns
+     *   Path patterns with '%s' placeholder for the entity id.
+     * @param bool $negate
+     *   FALSE, if $expected describes access being granted.
+     *   TRUE, if $expected describes access being denied.
+     */
+    protected function assertEntityPathsAccess(array $expected, array $accounts, array $entities, array $sprintf_path_patterns, bool $negate = FALSE) {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: Why this wrapper function? You can do the assertions related to paths here. Remember that tests are also documentation about the module, the more complex, the harder for other people to read and follow.

File: README.md [hunk 9]
```diff
@@ -197,24 +197,30 @@ Collabora is used within Drupal.  Most of the time this permission is
...
+  Users with this permission are allowed to preview documents attached
+  to an unpublished media entity of the given type, using the Collabora Online
+  editor in preview/readonly mode.
+
+In the current version of this module, the 'administer media' permission
+from Drupal core grants access to all media operations, including the use
+of the Collabora Online editor for preview and edit.
+
+For a consistent experience, it is recommended that a role with edit
+permissions should also be granted respective preview permissions.
```
Comment (AaronGilMartinez) [inline README.md]: Not sure about this last paragraph. Is up to the sites how they want to configure permissions.

File: collabora_online.module [hunk 10]
```diff
@@ -123,24 +123,59 @@ function collabora_online_entity_operation(EntityInterface $entity) {
  * Checks access for the new media operations provided by this module.
  */
 function collabora_online_media_access(MediaInterface $media, string $operation, AccountInterface $account): AccessResultInterface {
-    $type = $media->bundle();
```
Comment (AaronGilMartinez) [inline collabora_online.module:126]: Why this change of scope?

File: collabora_online.module [hunk 11]
```diff
@@ -123,24 +123,59 @@ function collabora_online_entity_operation(EntityInterface $entity) {
...
     switch ($operation) {
         case 'preview in collabora':
-            return AccessResult::allowedIfHasPermission($account, "preview $type in collabora");
+            $type = $media->bundle();
+            if ($media->isPublished()) {
+                return AccessResult::allowedIfHasPermission($account, "preview $type in collabora")
+                    ->addCacheableDependency($media);
+            }
+            $preview_own_permission = "preview own unpublished $type in collabora";
+            if (!$account->hasPermission($preview_own_permission)) {
```
Comment (AaronGilMartinez) [inline collabora_online.module]: Only reason I see to use neutral is to set 'reasons', if the user doesn't have the permission is handled by AccessResult::allowedIfHasPermission().

I think its a bit too much to place this conditions just to set some 'reasons' where the access is neutral when we only need to return allowed under certain conditions.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+            ],
+            $accounts,
+            $media_entities,
+            [
+                'preview in collabora',
+                'edit in collabora',
+            ],
+        );
+
+        $this->assertEntityPathsAccess(
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: Since routes check access with the permissions they are going to have exact same results that entity access. I feel like this is a duplication of access checks. What is the value of this path checks?

File: collabora_online.module [hunk 13]
```diff
@@ -123,24 +123,59 @@ function collabora_online_entity_operation(EntityInterface $entity) {
...
+            if ($is_owner) {
+                $access_result = AccessResult::allowed();
+            }
+            else {
+                $access_result = AccessResult::neutral()
+                    ->setReason("The user has the '$preview_own_permission' permission, but is not the owner of the media item.");
+            }
+            return $access_result
+                ->cachePerPermissions()
+                ->cachePerUser()
```
Comment (AaronGilMartinez) [inline collabora_online.module:149]: Why  user and permissions? user is more granular.

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

File: src/CollaboraMediaPermissions.php (hunk 2 above)
Comment (donquixote) [reply src/CollaboraMediaPermissions.php:80]: Yes absolutely.
We want to indicate that this permission does not grant access to unpublished media anymore.
This is also in line with similar permissions in core, e.g. "View published content".

File: collabora_online.module (hunk 11 above)
Comment (donquixote) [reply collabora_online.module]: Indeed, it's quite the overhead to add the reasons.

File: collabora_online.module (hunk 13 above)
Comment (donquixote) [reply collabora_online.module:149]: I think you are right.
If we add permissions to a role, does the access cache get reset for all users with that role?
Same if we add a role to a user?

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 12 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: True, it is duplication.
We want to verify that the urls are checked correctly, and that the entity access does map to the routes access.

We don't really need to test the full matrix for this, we could instead do some example urls.
But this would require me to think, whereas doing the same matrix again is a no-brainer solution..

And this test is cheap (fast), so the duplication is not hurting us much. The main problem would be the duplication in code, especially if we need to update it in the future.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 5 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php:157]: We need to verify that anonymous id = 0 is not treated as the owner of a document with id = 0.
We have this as a condition in the is_owner check.
`$is_owner = ($account->id() && $account->id() == $media->getOwnerId());`

Otherwise indeed we are testing too much here.
I also need to rethink the other cases above.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 12 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: Any thoughts on which permission combinations we should check with paths, specifically?
If we want sufficient coverage but no duplication.

Some options I can think of:
- Check a few permissions with paths, but not the full range.
- Check the same permissions with paths as with entity access calls, but hide/automate the duplication, e.g. in an assert method that checks paths and entity access using the same data.
- Implement a custom entity access hook in a test method, connected to a service that is remote-controlled from the test.
- Mock the access control handler for media.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 4 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: It was a leftover.
Removed.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 6 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: You are right.
Anyway I changed a lot now..

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 7 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: The motivation for the parameter was to have the assertion method be less opinionated, and instead have the opinionated part all in one place with the calling code.
If a method is called `assertEntityAccess()`, it should not have specific hard-coded operations that only apply to Collabora Online.
Anyway, it all changed now.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 8 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: Yeah, i was tempted to split out methods so that individual methods become smaller.
But I think now that in these test classes, the number of methods and the call stack size can be a bigger source of complexity than the length of an individual method.

File: collabora_online.module (hunk 11 above)
Comment (donquixote) [reply collabora_online.module]: Actually the reason is that if the permission is there, we then need to check for ownership.
What we _can_ do is introduce a local variable like:
```php
            $preview_own_access = AccessResult::allowedIfHasPermission($account, $preview_own_permission)
                ->addCacheableDependency($media);
            if (!$preview_own_access->isAllowed()) {
                return $preview_own_access;
            }
```
This way we don't need to add the custom reason string.

File: collabora_online.module (hunk 13 above)
Comment (donquixote) [reply collabora_online.module:149]: I just checked in core, in most places with `->cachePerUser()` there is no `->cachePerPermission()`.
So this seems the safe thing to do.

File: collabora_online.module (hunk 10 above)
Comment (donquixote) [reply collabora_online.module:126]: It's a matter of taste.
We don't need `$type` in the `default:` switch case, and I like the code to be more local, even at the cost of duplication.
But... mainly I think it was a leftover of a different approach I did in the past, when we were checking view access.
I can move it back.

Comment (donquixote) [reply collabora_online.module:126]: Moved back.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 12 above)
Comment (AaronGilMartinez) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: Usually is a dedicated test only for that, like this: https://git.drupalcode.org/project/file_defaults_by_type/-/blob/1.x/modules/file_defaults_by_type_ui/tests/src/Kernel/RouteAccessTest.php

That would mean duplicate code indeed, since you want to cover all the cases. Or you create some code to provide data or you do the second point.

Second option seems the more sensible for me since the code is there and ready.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 5 above)
Comment (AaronGilMartinez) [reply tests/src/Kernel/CollaboraMediaAccessTest.php:157]: Lets keep the new version of the test for anoymous, I would say that is testing the ownership but access is a sensitive topic for this module, and now the test has a more limited scope.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+        $this->createMediaType('file', ['id' => 'document']);
+        $this->createMediaType('file', ['id' => 'book']);
+
+        // Consume the user id 1.
+        $this->createUser();
+    }
+
+    /**
+     * Tests different permission combinations.
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php]: This is about access not the permissions combinations.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+        $actual_path_access = [];
+        foreach ($entities as $entity_key => $entity) {
+            foreach ($sprintf_path_patterns as $pattern_key => $sprintf_path_pattern) {
+                $path = sprintf($sprintf_path_pattern, $entity->id());
+                $has_path_access = Url::fromUserInput($path)->access($account);
+                if ($has_path_access) {
+                    $actual_path_access[$entity_key][] = $pattern_key;
+                }
+            }
+        }
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:267]: Actually, I think you don't need to duplicate loops, since you have to loop same entities and operations.
Just need to store results in the different variables and check after.

This would reduce the duplicated code and keep doing path checks.

And also have descriptive messages for each.

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

**Review by AaronGilMartinez — APPROVED** [review]
